feat(web-analytics): Allow customers to compare against a different period

rafaeelaudibert opened the pull request:
## Changes

Right now, we'd always compare our data against the last period, i.e. if we're looking at the data from the last 14 days, then you'd always compare it against the previous 14 days.

This is not always ideal, as someone might want to compare against numbers from the same time last year.

We've added the comparator component in a previous commit (hidden behind a FF) and we're now actually using that data in the backend.

![image](https://github.com/user-attachments/assets/bb22d1da-f281-454a-b23a-a932c04f2887)


## Does this work well for both Cloud and self-hosted?

Yep

## How did you test this code?

- [x] Manual UI tests 
- [x] Backend tests

rafaeelaudibert committed:
Stop displaying unrelated-to-conversion-goal sections

rafaeelaudibert committed:
feat(web-analytics): Allow customers to compare against a different period

Right now, we'd always compare our data against the last period, i.e. if we're looking at the data from the last 14 days, then you'd always compare it against the previous 14 days.

This is not always ideal, as someone might want to compare against numbers from the same time last year.

We've added the comparator component in a previous commit (hidden behind a FF) and we're now actually using that data in the backend.

rafaeelaudibert commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -882,6 +882,7 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
                                                   kind: NodeKind.WebExternalClicksTableQuery,
                                                   properties: webAnalyticsFilters,
                                                   dateRange,
+                                                  compareFilter, // NOTE: Backend is not using this yet
```
@robbie-c Thoughts on whether this is actually helpful? Does it make sense to compare outbound link clicks? The backend doesn't compare it at all, it's always just current data

It's not hard to implement, but I'm trying to get a feeling here on how this could be helpful. I think it is, but I wanna hear from you.

robbie-c replied:
Yeah I think it is, but fine if we don't include it in this PR

rafaeelaudibert replied:
I've included it :)

rafaeelaudibert committed:
refactor: Simplify `web_overview`

Use same methods as we're using on `stats_table` to calculate previous metrics. We've extracted this up to `web_analytics_query_runner`

rafaeelaudibert commented on frontend/__snapshots__/scenes-other-onboarding--onboarding-billing--light.png:
Come on buddy, why do you keep disappearing 😮‍💨

rafaeelaudibert committed:
Implement comparison for Outbound clicks

rafaeelaudibert committed:
test: Add missing tests for comparison

rafaeelaudibert committed:
fix: fix mypy error

rafaeelaudibert committed:
fix: Remove default of -7d for comparefilter

This is breaking some of our code and it's also a bad default

rafaeelaudibert committed:
fix: Regenerate mypy baseline

rafaeelaudibert committed:
test: Add new entry to insights default fields

robbie-c commented on posthog/hogql_queries/web_analytics/external_clicks.py:
```diff
@@ -43,36 +43,38 @@ def to_query(self) -> ast.SelectQuery:
 )
 GROUP BY "context.columns.url"
 ORDER BY "context.columns.visitors" DESC,
+"context.columns.clicks" DESC,
```
nice drive-by, makes sense

robbie-c commented on posthog/hogql_queries/web_analytics/external_clicks.py:
```diff
@@ -43,36 +43,38 @@ def to_query(self) -> ast.SelectQuery:
         count() AS filtered_click_count,
-        {url_expr} AS url
+        {url_expr} AS url,
+        MIN(timestamp) AS start_timestamp
```
This will be the timestamp of the first click rather than the start of the session. If you want the start of the session you can do `any(session.$start_timestamp) as start_timestamp`

rafaeelaudibert replied:
good catch

robbie-c commented on posthog/hogql_queries/utils/query_previous_period_date_range.py:
```diff
@@ -15,13 +15,13 @@ class QueryPreviousPeriodDateRange(QueryDateRange):
 
     _team: Team
-    _date_range: Optional[InsightDateRange]
+    _date_range: Optional[InsightDateRange | DateRange]
```
A: They look the same except for defaults
```python
class InsightDateRange(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
    )
    date_from: Optional[str] = "-7d"
    date_to: Optional[str] = None
    explicitDate: Optional[bool] = Field(
        default=False,
        description=(
            "Whether the date_from and date_to should be used verbatim. Disables rounding to the start and end of"
            " period."
        ),
    )

class DateRange(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
    )
    date_from: Optional[str] = None
    date_to: Optional[str] = None
    explicitDate: Optional[bool] = Field(
        default=False,
        description=(
            "Whether the date_from and date_to should be used verbatim. Disables rounding to the start and end of"
            " period."
        ),
    )
```

robbie-c replied:
Is there any reason not to use InsightDateRange?

rafaeelaudibert replied:
Basically because of the default. Insights want to set 7 days as default, we're happy with nothing. I'll change it to have no default to improve the code readability, we can always revert it

robbie-c reviewed (commented):
<img width="1246" alt="Screenshot 2024-12-14 at 15 56 07" src="https://github.com/user-attachments/assets/a703b715-1641-4666-bc92-18724c4e87c4" />

I got this error on the paths insight when I went from "no comparison" to "compare to previous period"

rafaeelaudibert commented on mypy-baseline.txt:
```diff
@@ -1,67 +1,4 @@
 posthog/warehouse/models/ssh_tunnel.py:0: error: Incompatible types in assignment (expression has type "NoEncryption", variable has type "BestAvailableEncryption")  [assignment]
-posthog/temporal/data_imports/pipelines/sql_database_v2/schema_types.py:0: error: Statement is unreachable  [unreachable]
```
There was a single difference. I really don't know why everything was reordered 🤔

robbie-c replied:
Yeah fair enough - don't worry too much about it. In general we should try to reduce this list but IMO we don't need to be too strict

rafaeelaudibert committed:
chore: Use `DateRange` everywhere instead of `InsightDateRange`

These are basically the same, the only difference is the default. Let's use `DateRange` without the default everywhere. If this proves to be a problem (too many broken tests) we can either:?
1. revert this
2. try to use it with the default on the places where we were not using a default before

fix: Remove last references to `InsightDateRange`

rafaeelaudibert committed:
fix: Use session timestamp rather than first click timestamp

rafaeelaudibert committed:
fix: Guarantee we wont break if called with empty number

rafaeelaudibert commented:
> I got this error on the paths insight when I went from "no comparison" to "compare to previous period"

@robbie-c I think this is happening because you have some cached data, I can't reproduce it 🤔.

I've added a failsafe to guarantee that doesn't happen, though.

rafaeelaudibert committed:
fix: fix broken test

robbie-c reviewed (approved):
:shipit:
